cmd/mtaresolver.go

```diff
@@ -0,0 +1,392 @@
+import (
+	"encoding/json"
+	"fmt"
+	"path"
+
+	"os"
+	"strings"
+
+	"github.com/SAP/cloud-mta/mta"
+	"github.com/joho/godotenv"
+)
+
+type MTAResolver struct {
+	mta.MTA
+	WorkingDir string
+	context    *ResolveContext
+}
+
+const RESOURCE_TYPE = 1
+const MODULE_TYPE = 2
+
+const VARIABLE_PREFIX = "~"
+const PLACEHOLDER_PREFIX = "$"
+const TEMPLATE_PREIX = "^"
+
+type MTASource struct {
+	Name       string
+	Parameters map[string]interface{} `yaml:"parameters,omitempty"`
+	Properties map[string]interface{} `yaml:"properties,omitempty"`
+	Type       int
+	Module     *mta.Module
+	Resource   *mta.Resource
+}
+
+type EnvVariables map[string]interface{}
+
+type Package struct {
+	Main    string                 `json:"main"`
+	Scripts map[string]interface{} `json:"scripts"`
+}
+
+func NewMTAResolver(m *mta.MTA, workspaceDir string) *MTAResolver {
+	resolver := &MTAResolver{*m, workspaceDir, &ResolveContext{
+		global:    map[string]string{},
+		modules:   map[string]map[string]string{},
+		resources: map[string]map[string]string{},
+	}}
+
+	for _, module := range m.Modules {
+		resolver.context.modules[module.Name] = map[string]string{}
+	}
+
+	for _, resource := range m.Resources {
+		resolver.context.resources[resource.Name] = map[string]string{}
+	}
+	return resolver
+}
+
+func (s *MTASource) IsResource() bool {
+	return s.Type == RESOURCE_TYPE
+}
+
+func (m *MTAResolver) ResolveProperies(module *mta.Module) {
+
+	if m.Parameters == nil {
+		m.Parameters = map[string]interface{}{}
+	}
+
+	//add env variables
+	for _, val := range os.Environ() {
+		pos := strings.Index(val, "=")
+		if pos > 0 {
+			key := strings.Trim(val[:pos], " ")
+			value := strings.Trim(val[pos+1:], " ")
+			m.addValueToContext(key, value)
+		}
+	}
+
+	//add .env file in module's path to the module context
+	if len(module.Path) > 0 {
+		envFile := path.Join(m.WorkingDir, module.Path, ".env")
+		envMap, err := godotenv.Read(envFile)
+		if err == nil {
+			for key, value := range envMap {
+				m.addValueToContext(key, value)
+			}
+		}
+	}
+	m.addServiceNames(module)
+
+	//top level properties
+	for key, value := range module.Properties {
+		//no expected variables
+		propValue := m.resolve(module, nil, value)
+		module.Properties[key] = m.resolvePlaceholders(module, nil, nil, propValue)
+	}
+
+	//required properties:
+	for _, req := range module.Requires {
+		requiredSource := m.FindProvider(req.Name)
+		for propName, PropValue := range req.Properties {
+			resolvedValue := m.resolve(module, &req, PropValue)
+			//replace value with resolved value
+			req.Properties[propName] = m.resolvePlaceholders(module, requiredSource, &req, resolvedValue)
+		}
+	}
+}
+
+func (m *MTAResolver) addValueToContext(key, value string) {
+	//if the key has format of "module/key", or "resource/key" writes the value to the module's context
+	slashPos := strings.Index(key, "/")
+	if slashPos > 0 {
+		modName := key[:slashPos]
+		key = key[slashPos+1:]
+		modContext, ok := m.context.modules[modName]
```

Comment:
it's worth to rename modContext to (for example) context

Reply:
changed to modulesContext. context may confuse with m.context